google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkServiceManager.java

```diff
@@ -43,6 +77,196 @@ public void onNewCloudSdkServiceTypeSelected(CloudSdkServiceType newServiceType)
     }
   }
 
+  /**
+   * Waits in background for Cloud SDK to be ready for all operations and then runs the given
+   * runnable. If process results in error or user cancel, shows notification and does not run. This
+   * method does not block and must be called from application UI thread.
+   *
+   * @param project Project to which runnable belongs.
+   * @param runAfterSdkReady Runnable to execute after Cloud SDK is ready. This runnable will be
+   *     executed on the application UI thread.
+   * @param progressMessage Message to show in progress dialog to identify which process is
+   *     starting, i.e. deployment or local server.
+   * @param sdkLogger Logger for errors etc.
+   */
+  public void runWhenSdkReady(
+      @Nullable Project project,
+      @NotNull Runnable runAfterSdkReady,
+      String progressMessage,
+      CloudSdkLogger sdkLogger) {
+    doWait(
+        project,
+        () -> {
+          // at this point the installation should be either ready, failed or user cancelled.
+          // run only if ready.
+          if (CloudSdkService.getInstance().getStatus() == SdkStatus.READY) {
+            ApplicationManager.getApplication().invokeLater(runAfterSdkReady);
+          }
+        },
+        progressMessage,
+        sdkLogger);
+  }
+
+  /**
+   * Blocks current thread until Cloud SDK is ready for all operations. If process results in error
+   * or user cancel, calls {@link CloudSdkLogger} methods to notify about errors and shows
+   * notifications. This method is expected to be called from non-UI background thread.
+   *
+   * @param project Project to which SDK operation belongs.
+   * @param progressMessage Message to show in progress dialog to identify which process is
+   *     starting, i.e. deployment or local server.
+   * @param sdkLogger Callback to log errors and progress.
+   */
+  public void waitWhenSdkReady(
+      @Nullable Project project, String progressMessage, CloudSdkLogger sdkLogger)
+      throws InterruptedException {
+    CountDownLatch blockingCompletedLatch = new CountDownLatch(1);
+    ApplicationManager.getApplication()
+        .invokeLater(
+            () -> doWait(project, blockingCompletedLatch::countDown, progressMessage, sdkLogger));
+    blockingCompletedLatch.await();
+  }
+
+  /**
+   * Performs generic wait in a separate thread for SDK to become ready, returns immediately. Must
+   * be called from UI thread.
+   */
+  private void doWait(
```

Comment:
nit, but this method feels a bit long to be able to easily digest. is it possible to break its contents down into well-named methods?

Reply:
Sure, thinking about it.